File: 2024/Jour11/day11.py
"""
Day 11 of advent of code
Too slow in R so trying memoization in Python
"""

import functools

Chemin = "C:/Users/gmulier/Documents/Github/AdventCode/2024"
Input = open(Chemin + "/Input/day11.txt").read().strip()
Input = Input.split(" ")

# Part 1

@functools.cache
def BlinkRecurr (Pierre, IterRest):
    """
    Recursive function to compute number of stones after the iterations

    :param Pierre: Number written on stone
    :param IterRest: Number of iteration left
    :return: Number of stone after the IterRest iterations
    """
    if IterRest == 0:
        return 1
    elif Pierre == "0":
        return BlinkRecurr("1", IterRest - 1)
    elif (len(Pierre) % 2) == 0:
        MoitieUne = BlinkRecurr(str(int(Pierre[0:(len(Pierre) // 2)])), IterRest - 1)
        MoitieDeux = BlinkRecurr(str(int(Pierre[(len(Pierre)//2):len(Pierre)])), IterRest - 1)
        return MoitieUne + MoitieDeux
    else:
        return BlinkRecurr(str(int(Pierre) * 2024), IterRest - 1)

ResPart1 = 0
# Loop over the Input and update the result with the function
for s in Input:
    ResPart1 += BlinkRecurr(s, 25)
print("Résultat part 1 : " + str(ResPart1))

# Part 2 : memoization
ResPart2 = 0
for s in Input:
    ResPart2 += BlinkRecurr(s, 75)
print("Résultat part 2 : " + str(ResPart2))

File: 2024/Jour11/day11.R
# Day 11 advent of code 2024 #

# Part 1 ----

library(purrr)

Chemin <- "C:/Users/gmulier/Documents/Github/AdventCode/2024"
Input <- readLines(paste0(Chemin, "/Input/day11.txt"))
# Input <- "125 17"
Input <- strsplit(Input, " ")[[1]]

Blink <- function(String) {
  if (as.numeric(String) == 0) {
    return("1")
  } else if (nchar(String) %% 2 == 0) {
    return(as.character(as.numeric(c(substr(String, 1, nchar(String) %/% 2), substr(String, nchar(String) %/% 2 + 1, nchar(String))))))
  } else {
    return(as.character(2024 * as.numeric(String)))
  }
}
Res <- reduce(1:25, \(x, y) unlist(lapply(x, Blink)), .init = Input)
print(paste0("Réponse partie 1 : ", length(Res)))


# Part 2 ----

# Trying memoization
library(memoise)
BlinkRecur <- function(Pierre, IterRest) {
  if (IterRest == 0) {
    return(1)
  } else if (Pierre == "0") {
    return(BlinkRecur("1", IterRest - 1))
    # return(2)
  } else if (nchar(Pierre) %% 2 == 0) {
    # return(3)
    MoitieUne <- BlinkRecur(as.character(as.numeric(substr(Pierre, 1, nchar(Pierre) %/% 2))), IterRest - 1)
    MoitieDeux <- BlinkRecur(as.character(as.numeric(substr(Pierre, 1 + nchar(Pierre) %/% 2, nchar(Pierre)))), IterRest - 1)
    return(MoitieUne + MoitieDeux)
  } else {
    # return(4)
    return(BlinkRecur(as.character(as.numeric(Pierre) * 2024), IterRest - 1))
  }
}
BlinkRecurMem <- memoise(BlinkRecur)
Res <- sum(vapply(Input, \(x) {
  print(paste0("Item n°", match(x, Input), " : ", x))
  BlinkRecurMem(x, 75)
}, numeric(1)))
print(paste0("Réponse partie 2 : ", Res))
# Impossible to finish. I translated it in python

